File: 2. Array and String/7. intersection/intersection.py
# def intersection(a, b):
#   setA = set(a)
#   setB = set(b)
#   return list(setA & setB)

def intersection(a, b):
  set_a = set(a)
  return [ num for num in b if num in set_a]

def intersection1(a, b):
  set_a = set(a)
  set_b = set(b)
  return list(set_a & set_b)

print(intersection([4,2,1,6], [3,6,9,2,10]))
print(intersection([2,4,6], [4,2]))
print(intersection([4,2,1], [1,2,4,6]))
print(intersection([0,1,2], [10,11]))
print(intersection1([4,2,1,6], [3,6,9,2,10]))
print(intersection1([2,4,6], [4,2]))
print(intersection1([4,2,1], [1,2,4,6]))
print(intersection1([0,1,2], [10,11]))

a = [ i for i in range(0, 50000) ]
b = [ i for i in range(0, 50000) ]
print(intersection(a,b))
print(intersection1(a,b))

File: 2. Array and String/7. intersection/intersection.js

// const intersection = (a, b) => {
//     return a.filter((aElement) => {
//         return b.includes(aElement);
//     })
// }

// Complexity
// n = length of array a, m = length of array b
// Time: O(n*m)
// Space: O(min(n,m))

// When the array contains duplicate elements
// const intersection = (a, b) => {
//     if(a.length > b.length)
//         [a, b] = [b, a];

//     let aMap = {};
//     for(let i=0; i<a.length; i++){
//         if(aMap[a[i]] === undefined)
//             aMap[a[i]] = 0;
//         aMap[a[i]]++;
//     }

//     let intersection = [];
//     for(let num of b){
//         if(aMap[num] !== undefined){
//             intersection.push(num);
//         }
//     }

//     return intersection;
// }


// When the array does not contain dupliate elements
const intersection = (a, b) => {
  const result = [];
  const setA = new Set(a);
  for (let item of b) {
    if (setA.has(item)) {
      result.push(item);
    }
  }
  return result;
};

// Complexity
// n = length of array a, m = length of array b
// Time: O(n+m)
// Space: O(n)

// let ans = intersection([4,2,1,6], [3,6,9,2,10]);
// console.log(`Ans: ${ans}`);